XbSymbolDatabase.c

```diff
@@ -359,21 +488,190 @@ uint32_t XbSymbolLibraryToFlag(const char* library_name)
+    const xbe_section_header* xb_section_headers = (xbe_section_header*)(xb_start_addr + pXbeHeader->pSectionHeadersAddr);
+    unsigned int section_total = pXbeHeader->dwSections;
+    unsigned int section_index = 0;
+    unsigned int count = 0;
+    uint16_t build_version = 0;
+    bool has_dsound_library = false;
+    xb_xbe_type XbeType = GetXbeType(xb_header_addr);
+    OutputHandler output;
+    output.func = g_output_func;
+    output.verbose_level = g_output_verbose_level;
+
+    // Only process XDK applications.
+    if (pXbeHeader->pLibraryVersionsAddr != 0) {
+
+        for (library_index; library_index < library_total; library_index++) {
+
+            library_flag = XbSymbolDatabase_LibraryToFlag(xb_library_versions[library_index].szName);
+
+            // If library is unknown to the database, skip it.
+            if (library_flag == 0) {
+                continue;
+            }
+
+            // Keep the highest build version for manual checklist.
+            if (build_version < xb_library_versions[library_index].wBuildVersion) {
+                build_version = xb_library_versions[library_index].wBuildVersion;
+            }
+
+            // If found DSOUND library, then skip the manual check.
+            if (library_flag == XbSymbolLib_DSOUND && !has_dsound_library) {
+                has_dsound_library = true;
+            }
+
+            // Append the information to the array.
+            if (library_header != NULL) {
+
+                library_header->filters[count].flag = library_flag;
+                library_header->filters[count].build_version = xb_library_versions[library_index].wBuildVersion;
+                library_header->filters[count].qfe_version = xb_library_versions[library_index].wFlags.QFEVersion;
+                memcpy(library_header->filters[count].name, XbSymbolDatabase_LibraryToString(library_flag), 8);
+
+                if ((library_flag & (XbSymbolLib_D3D8 | XbSymbolLib_D3D8LTCG)) > 0) {
+
+                    // Functions in this library were updated by June 2003 XDK (5558) with Integrated Hotfixes,
+                    // However August 2003 XDK (5659) still uses the old function.
+                    // Please use updated 5788 instead.
+                    if (xb_library_versions[library_index].wBuildVersion >= 5558 &&
+                        xb_library_versions[library_index].wBuildVersion <= 5659 &&
+                        xb_library_versions[library_index].wFlags.QFEVersion > 1) {
+
+                        output_message_format(&output, XB_OUTPUT_MESSAGE_WARN,
+                            "D3D8 version 1.0.%d.%d Title Detected: This game uses an alias version 1.0.5788",
+                            xb_library_versions[library_index].wBuildVersion,
+                            xb_library_versions[library_index].wFlags.QFEVersion);
+
+                        library_header->filters[count].build_version = 5788;
+                    }
+                }
+            }
+            count++;
```

Comment:
This causes the `filters` array to have gaps. If that's undesirable, increase `count` only when populating a filter.

Same happens below, I'll mark them with a single `gap` comment.

Reply:
I think I see what you mean, I'll re-review this function again.

Reply:
I don't see the problem. I think you had missed the API's documentation. If `library_header` is null pointer, it will return the counts for what is needed to allocate filter's array. Then re-call the function again to fill in the filters.